src/cplscheme/BiCouplingScheme.cpp

```diff
@@ -138,23 +138,67 @@ CouplingData *BiCouplingScheme::getReceiveData(
 void BiCouplingScheme::exchangeInitialData()
 {
   // F: send, receive, S: receive, send
+  bool recvInitialData = true;
+  bool sendInitialData = true;
+
   if (doesFirstStep()) {
     if (sendsInitializedData()) {
-      sendData(getM2N(), getSendData());
+      sendData(getM2N(), getSendData(), sendInitialData);
     }
     if (receivesInitializedData()) {
-      receiveData(getM2N(), getReceiveData());
+      receiveData(getM2N(), getReceiveData(), recvInitialData);
       checkDataHasBeenReceived();
+    } else {
+      initializeZeroReceiveData(getReceiveData());
     }
   } else { // second participant
     if (receivesInitializedData()) {
-      receiveData(getM2N(), getReceiveData());
+      receiveData(getM2N(), getReceiveData(), recvInitialData);
       checkDataHasBeenReceived();
+    } else {
+      initializeZeroReceiveData(getReceiveData());
     }
     if (sendsInitializedData()) {
-      sendData(getM2N(), getSendData());
+      sendData(getM2N(), getSendData(), sendInitialData);
+    }
+  }
+}
+
+void BiCouplingScheme::storeTimeStepReceiveData(double relativeDt)
+{
+  PRECICE_ASSERT(math::greaterEquals(relativeDt, time::Storage::WINDOW_START), relativeDt);
+  PRECICE_ASSERT(math::greaterEquals(time::Storage::WINDOW_END, relativeDt), relativeDt);
+  if (hasDataBeenReceived()) {
```

Comment:
Wouldn't it be easier to understand if this `if` was outside of this function?

Reply:
I think this `if` is not even needed. Will try to remove it.